met-api/src/met_api/services/widget_translation_service.py

```diff
@@ -0,0 +1,135 @@
+"""Service for widget translation management."""
+from http import HTTPStatus
+
+from sqlalchemy.exc import IntegrityError
+from met_api.constants.membership_type import MembershipType
+from met_api.constants.widget import WidgetType
+from met_api.exceptions.business_exception import BusinessException
+from met_api.models.language import Language as LanguageModel
+from met_api.models.widget import Widget as WidgetModel
+from met_api.models.widget_map import WidgetMap as WidgetMapModel
+from met_api.models.widget_poll import Poll as PollModel
+from met_api.models.widget_translation import WidgetTranslation as WidgetTranslationModel
+from met_api.models.widget_video import WidgetVideo as WidgetVideoModel
+from met_api.schemas.widget_translation import WidgetTranslationSchema
+from met_api.services import authorization
+from met_api.utils.roles import Role
+
+
+class WidgetTranslationService:
+    """Widget translation management service."""
+
+    @staticmethod
+    def get_translation_by_widget_id_and_language_id(widget_id=None, language_id=None):
+        """Get translation by widget id and language id."""
+        widget_translation_schema = WidgetTranslationSchema(many=True)
+        widgets_translation_records =\
+            WidgetTranslationModel.get_translation_by_widget_id_and_language_id(widget_id, language_id)
+        widget_translations = widget_translation_schema.dump(widgets_translation_records)
+        return widget_translations
+
+    @staticmethod
+    def create_widget_translation(translation_data, pre_populate=True):
+        """Create widget translation item."""
+        try:
+            widget = WidgetModel.find_by_id(translation_data['widget_id'])
+            if not widget:
+                raise ValueError('Widget to translate was not found')
+
+            one_of_roles = (
+                MembershipType.TEAM_MEMBER.name,
+                Role.EDIT_ENGAGEMENT.value
+            )
+            authorization.check_auth(one_of_roles=one_of_roles, engagement_id=widget.engagement_id)
+
+            language_record = LanguageModel.find_by_id(translation_data['language_id'])
+            if not language_record:
+                raise ValueError('Language to translate was not found')
+
+            if pre_populate:
+                # prepopulate translation with base language data
+                WidgetTranslationService._get_default_language_values(widget, translation_data)
+
+            created_widget_translation = WidgetTranslationModel.create_widget_translation(translation_data)
+            return WidgetTranslationSchema().dump(created_widget_translation)
+        except IntegrityError as e:
```

Comment:
Will this be able to catch when we have two translations for the same widget? The widget_id and language_id combination needs to be unique

Reply:
Yes I tried to post two requests with the same widget_id and language_id, the second time it gives a BusinessException